auth/user_mgt_test.go

```diff
@@ -0,0 +1,716 @@
+		}, {
+			(&UserToCreate{}).Email(""),
+			`invalid Email: "" Email must be a non-empty string`,
+		}, {
+			(&UserToCreate{}).Email("a"),
+			`malformed email address string: "a"`,
+		}, {
+			(&UserToCreate{}).Email("a@"),
+			`malformed email address string: "a@"`,
+		}, {
+			(&UserToCreate{}).Email("@a"),
+			`malformed email address string: "@a"`,
+		}, {
+			(&UserToCreate{}).Email("a@a@a"),
+			`malformed email address string: "a@a@a"`,
+		},
+	}
+	for i, test := range badUserParams {
+		_, err := client.CreateUser(context.Background(), test.params)
+		if err == nil {
+			t.Errorf("%d) got no error, wanted error %s", i, test.expectingError)
+		}
+		if err.Error() != test.expectingError {
+			t.Errorf(`got error: "%s" wanted error: "%s"`, err.Error(), test.expectingError)
+		}
+	}
+
+}
+
+func TestCreateUser(t *testing.T) {
+	s, closer := echoServer([]byte(`{
+		"kind": "identitytoolkit#SignupNewUserResponse",
+		"email": "",
+		"localId": "expectedUserID"
+	   }`), t)
+	defer closer()
+	goodParams := []*UserToCreate{
+		nil,
+		{},
+		(&UserToCreate{}).Password("123456"),
+		(&UserToCreate{}).UID("1"),
+		(&UserToCreate{}).UID(strings.Repeat("a", 128)),
+		(&UserToCreate{}).PhoneNumber("+1"),
+		(&UserToCreate{}).DisplayName("a"),
+		(&UserToCreate{}).Email("a@a"),
+		(&UserToCreate{}).PhoneNumber("+1"),
+	}
+	for _, par := range goodParams {
+		_, err := s.Client.CreateUser(context.Background(), par)
+		// There are two calls to the server, the first one, on creation retunrs the above []byte
+		// that's how we know the params passed validation
+		// the second call to GetUser, tries to get the user with the returned ID above, it fails
+		// with the following expected error
+		if err.Error() != "cannot find user map[localId:[expectedUserID]]" {
+			t.Error(err)
+		}
+	}
+}
+
+func TestBadUpdateParams(t *testing.T) {
+	badParams := []struct {
+		params         *UserToUpdate
+		expectingError string
+	}{
+		{
+			nil,
+			"params must not be empty for update",
+		}, {
+			&UserToUpdate{},
+			"params must not be empty for update",
+		}, {
+			(&UserToUpdate{}).CustomClaims(map[string]interface{}{"a": strings.Repeat("a", 993)}),
+			fmt.Sprintf(`Custom Claims payload must not exceed %d characters`, maxLenPayloadCC),
+		},
+	}
+
+	for _, res := range reservedClaims {
+		badParams = append(badParams,
+			struct {
+				params         *UserToUpdate
+				expectingError string
+			}{
+				(&UserToUpdate{}).CustomClaims(map[string]interface{}{res: true}),
+				fmt.Sprintf(`claim "%s" is reserved, and must not be set`, res)})
+	}
+
+	for i, test := range badParams {
+		_, err := client.UpdateUser(context.Background(), "outofstruct", test.params)
+		if err == nil {
+			t.Errorf("%d) got no error wanted error %s", i, test.expectingError)
+		}
+		if err.Error() != test.expectingError {
+			t.Errorf(`%d) got error "%s" wanted error "%s"`, i, err.Error(), test.expectingError)
+		}
+	}
+}
+
+func TestUpdateUser(t *testing.T) {
```

Comment:
Need test cases for cases where the backend server returns an error (e.g. 500 internal server error)